client/container_create.go

```diff
@@ -23,13 +23,20 @@ type configWrapper struct {
 func (cli *Client) ContainerCreate(ctx context.Context, config *container.Config, hostConfig *container.HostConfig, networkingConfig *network.NetworkingConfig, platform *ocispec.Platform, containerName string) (container.CreateResponse, error) {
 	var response container.CreateResponse
 
-	if err := cli.NewVersionError("1.25", "stop timeout"); config != nil && config.StopTimeout != nil && err != nil {
+	// Make sure we negotiated (if the client is configured to do so),
+	// as code below contains API-version specific handling of options.
+	//
+	// Normally, version-negotiation (if enabled) would not happen until
+	// the API request is made.
+	cli.checkVersion(ctx)
```

Comment:
I think this call is unneeded since the calls to `NewVersionError` below will do it anyway. But I'm not sure how go compiler order operations (ie. does it call `NewVersionError` first and check all conditions after?) :thinking: 

Having to call `cli.checkVersion` whenever `versions.*` functions are used feels a bit weird to me, IMO it's error-prone. I think the client should have its own helper functions wrapping `versions.*` to automatically call `cli.checkVersion(ctx)`.

Reply:
Yes, this one is indeed somewhat redundant (I probably should've called that out here), because `cli.NewVersionError` would trigger this as well.

I decided to put it here for visibility, and because I didn't want this to depend on the implicit side-effect of `cli.NewVersionError` doing the lookup (if the `cli.NewVersionError` lines would ever be removed, or moved _after_ the other checks, we would potentially introduce a regression).

> Having to call `cli.checkVersion` whenever `versions.*` functions are used feels a bit weird to me

Yes, I fully agree, and I don't really like this PR for that. My goal in this PR was to fix the immediate issue (which I think _is real_), and making the fix somewhat "back portable", and move from there.


We need to look at a good approach to;

- Allow API version negotiation
- Have (keep) it as "lazy" as possible (we had issues in the past where _local_ operations caused an API call to be made, which for sure isn't desirable)
- Allow concurrency (or at least, make things more "per request") for situations where the client is used as part of a longer-lived process (many places in the current design were made based on the assumptions that it would be used as part of the CLI, and that to be a short-lived process (run a command, and exit))


I have a follow-up / separate PR pending that would allow us to use the context to communicate the version (but that PR in itself is not yet tying into that; only moving the code so that we _CAN_ consider doing that); reviews on that are definitely welcome as well 😄 

- https://github.com/moby/moby/pull/46464

Reply:
> Yes, I fully agree, and I don't really like this PR for that. My goal in this PR was to fix the immediate issue (which I think is real), and making the fix somewhat "back portable", and move from there.

I'm fine with this PR as it's addressing an immediate issue we have :wink: But yeah, we should look into all the issues we have here.